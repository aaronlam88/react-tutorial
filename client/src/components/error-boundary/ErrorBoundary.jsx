import React, { Component } from 'react';
import PropTypes from 'prop-types';

class ErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError() {
    // Update state so the next render will show the fallback UI.
    return { hasError: true };
  }

  componentDidCatch(error, info) {
    // You can also log the error to an error reporting service
    console.error(error, info);
  }

  render() {
    const { hasError } = this.state;
    const { children } = this.props;
    if (hasError) {
      // You can render any custom fallback UI
      return <h1>Something went wrong.</h1>;
    }

    return children;
  }
}

ErrorBoundary.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.node,
  ]),
};

ErrorBoundary.defaultProps = {
  children: {},
};

export default ErrorBoundary;
