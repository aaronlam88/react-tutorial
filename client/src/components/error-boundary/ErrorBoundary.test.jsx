import toJson from "enzyme-to-json";
import { shallow } from "enzyme";

import React from "react";
import ErrorBoundary from "./ErrorBoundary";

it("should render children when no error", () => {
  const wrapper = shallow(<ErrorBoundary>Test</ErrorBoundary>);
  expect(toJson(wrapper)).toMatchSnapshot();
});

it("should render throw error", () => {
  const wrapper = shallow(<ErrorBoundary>Test</ErrorBoundary>);
  const instance = wrapper.instance();
  instance.constructor.getDerivedStateFromError();
  wrapper.setState({ hasError: true });
  expect(wrapper.contains(<h1>Something went wrong.</h1>)).toBeTruthy();
});

it("should log error to console", () => {
  global.console = { error: jest.fn() };
  const wrapper = shallow(<ErrorBoundary>Test</ErrorBoundary>);
  const instance = wrapper.instance();
  instance.componentDidCatch("error", "test");
  expect(console.error).toHaveBeenCalled();
});
