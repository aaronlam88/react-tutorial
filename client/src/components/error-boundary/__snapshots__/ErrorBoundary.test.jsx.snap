// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render children when no error 1`] = `"Test"`;
